customer/test/test_customer_service.py

```diff
@@ -0,0 +1,55 @@
+from unittest.mock import create_autospec
+
+import pytest
+
+from customer.domain.customer import CustomerCreate
+from customer.infra.customer_event_service import CustomerEventService
+from customer.infra.customer_repository import CustomerRepository
+from customer.service_layer.customer import CustomerService
+
+
+@pytest.fixture
+def customer_repository():
+    return create_autospec(CustomerRepository, instance=True)
+
+
+@pytest.fixture
+def customer_event_service():
+    return create_autospec(CustomerEventService, instance=True)
+
+
+@pytest.fixture
+def customer_service(
+    customer_repository,
+    customer_event_service,
+) -> CustomerService:
+    return CustomerService(
+        customer_repository=customer_repository,
+        customer_event_service=customer_event_service,
+    )
+
+
+class TestCustomerService:
+    def test_create(
+        self,
+        customer_service,
+        customer_repository,
+        customer_event_service,
+    ):
+        customer_service.create_customer(
+            first_name="Test",
+            last_name="Customer",
+            email="test@fake.com",
+        )
+
+        customer_repository.create.assert_called_once_with(
+            CustomerCreate(
```

Comment:
There's a bit of magic going on here. The way the `assert_called_x` functions work is to check that the arguments specified equal the arguments of the actual call in our code. So if we've created a new instance of `CustomerCreate` here, why does it match?

The answer is that it uses a special method on the objects to compare for equality: `__eq__()`. This is a special class of method on objects called "dunder" methods (i.e. double-underscore) which have special meanings in Python, and have a default definition for all classes (and of course, behing the scenes EVERYTHING in Python is an object!)

`CustomerCreate` subclasses `pydantic.BaseModel`, which defines it's `__eq__` as "if the other object has the same type and fields as this object, and all of the field values match, then it's equal. So in short, it recognises that we've passed the equivalent here and so matches the call.